### PR TITLE
Align React imports

### PR DESCRIPTION
**Pull request by kachkaev**
Follows #426 

- `FC`, `VFC`, `VoidFunctionComponent` → `FunctionComponent`
- `import React from "react"` & `import * as React from "react"` → `import { ... } from "react"`
- see PR comments

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1202583268528897

### CODE REVIEW

**Comment by nathggns** [comment]
Looking at this now

File: packages/@blockprotocol/graph/README.md [hunk 1]
```diff
@@ -59,8 +59,6 @@ const graphService = new GraphEmbedderHandler({
 For React embedding applications, we provide a `useGraphEmbedderService` hook, which accepts a `ref` to an element, and optionally any additional constructor arguments you wish to pass.
 
 ```javascript
-import React from "react";
-
```
Comment (nathggns) [inline packages/@blockprotocol/graph/README.md:63]: Other parts of this example use `React.` btw – so they probably need updating

File: packages/@blockprotocol/graph/tsconfig.json [hunk 2]
```diff
@@ -6,7 +6,7 @@
     "esModuleInterop": true,
     "forceConsistentCasingInFileNames": true,
     "isolatedModules": true,
-    "jsx": "preserve",
+    "jsx": "react-jsx",
```
Comment (nathggns) [inline packages/@blockprotocol/graph/tsconfig.json:9]: As above RE typescript/JSX.

File: packages/block-template/templates/custom-element/src/dev.tsx [hunk 3]
```diff
@@ -3,7 +3,6 @@
  * This file is not bundled with the block during the build process.
  */
 import { MockBlockDock } from "mock-block-dock";
-import React from "react";
 import ReactDOM from "react-dom";
```
Comment (nathggns) [inline packages/block-template/templates/custom-element/src/dev.tsx:6]: We probably ought to apply the same to `react-dom`, right?

File: packages/block-template/templates/custom-element/tsconfig.json [hunk 4]
```diff
@@ -4,7 +4,7 @@
     "allowSyntheticDefaultImports": true,
     "experimentalDecorators": true,
     "isolatedModules": true,
-    "jsx": "preserve",
+    "jsx": "react-jsx",
```
Comment (nathggns) [inline packages/block-template/templates/custom-element/tsconfig.json:7]: As above RE typescript/JSX.

File: packages/block-template/templates/react/tsconfig.json [hunk 5]
```diff
@@ -3,7 +3,7 @@
     "allowJs": true,
     "allowSyntheticDefaultImports": true,
     "isolatedModules": true,
-    "jsx": "preserve",
+    "jsx": "react-jsx",
```
Comment (nathggns) [inline packages/block-template/templates/react/tsconfig.json:6]: As above RE typescript/JSX.

File: packages/blockprotocol/tsconfig.json [hunk 6]
```diff
@@ -2,7 +2,7 @@
   "compilerOptions": {
     "allowJs": true,
     "isolatedModules": true,
-    "jsx": "preserve",
+    "jsx": "react-jsx",
```
Comment (nathggns) [inline packages/blockprotocol/tsconfig.json:5]: As above RE typescript/JSX.

File: site/src/components/pages/docs/search/search-suggested-links.tsx [hunk 7]
```diff
@@ -32,7 +31,7 @@ const SUGGESTED_LINKS: SuggestedLink[] = [
   },
 ];
 
-const SearchSuggestedLinks: React.VoidFunctionComponent<
+const SearchSuggestedLinks: React.FunctionComponent<
```
Comment (nathggns) [inline site/src/components/pages/docs/search/search-suggested-links.tsx:34]:
```suggestion
const SearchSuggestedLinks: FunctionComponent<
```

File: site/src/components/pages/home/why-block-protocol2section.tsx [hunk 8]
```diff
@@ -15,7 +14,7 @@ type WhyBlockProtocol2SectionProps = {
   fadeRef: React.Ref<HTMLDivElement>;
 };
 
-export const WhyBlockProtocol2Section: React.FC<
+export const WhyBlockProtocol2Section: React.FunctionComponent<
```
Comment (nathggns) [inline site/src/components/pages/home/why-block-protocol2section.tsx:17]:
```suggestion
export const WhyBlockProtocol2Section: FunctionComponent<
```

File: site/src/components/popover.tsx [hunk 9]
```diff
@@ -1,12 +1,11 @@
 /* eslint-disable-next-line -- allow import of original popover to extend it */
 import { Popover as MuiPopover, PopoverProps } from "@mui/material";
-import React from "react";
 import { useScrollLock } from "../util/mui-utils";
 
 /**
  * Custom Popover re-implementing MUI's troublesome scroll-lock mechanism.
  */
-export const Popover: React.FC<PopoverProps> = ({
+export const Popover: React.FunctionComponent<PopoverProps> = ({
```
Comment (nathggns) [inline site/src/components/popover.tsx:8]:
```suggestion
export const Popover: FunctionComponent<PopoverProps> = ({
```

File: site/src/components/service-message-list.tsx [hunk 10]
```diff
@@ -3,12 +3,12 @@ import {
...
 } from "@blockprotocol/core";
 import { Box, Typography } from "@mui/material";
-import * as React from "react";
+import { Fragment } from "react";
 
 import { JsonSchema } from "../lib/json-schema";
 import { mdxComponents } from "./mdx/mdx-components";
 
-const DataType: React.VoidFunctionComponent<{ propertySchema: JsonSchema }> = ({
+const DataType: React.FunctionComponent<{ propertySchema: JsonSchema }> = ({
```
Comment (nathggns) [inline site/src/components/service-message-list.tsx:11]:
```suggestion
const DataType: FunctionComponent<{ propertySchema: JsonSchema }> = ({
```

File: site/src/components/service-message-list.tsx [hunk 11]
```diff
@@ -31,7 +31,7 @@ const DataType: React.VoidFunctionComponent<{ propertySchema: JsonSchema }> = ({
   );
 };
 
-const ServiceMessageData: React.VoidFunctionComponent<{
+const ServiceMessageData: React.FunctionComponent<{
```
Comment (nathggns) [inline site/src/components/service-message-list.tsx:34]:
```suggestion
const ServiceMessageData: FunctionComponent<{
```

File: site/src/components/service-message-list.tsx [hunk 12]
```diff
@@ -100,7 +100,7 @@ const ServiceMessageData: React.VoidFunctionComponent<{
   );
 };
 
-const ServiceMessage: React.VoidFunctionComponent<{
+const ServiceMessage: React.FunctionComponent<{
```
Comment (nathggns) [inline site/src/components/service-message-list.tsx:103]:
```suggestion
const ServiceMessage: FunctionComponent<{
```

File: site/src/components/service-message-list.tsx [hunk 13]
```diff
@@ -169,7 +169,7 @@ const ServiceMessage: React.VoidFunctionComponent<{
   );
 };
 
-export const ServiceMessageList: React.VoidFunctionComponent<{
+export const ServiceMessageList: React.FunctionComponent<{
```
Comment (nathggns) [inline site/src/components/service-message-list.tsx:172]:
```suggestion
export const ServiceMessageList: FunctionComponent<{
```

File: site/tsconfig.json [hunk 14]
```diff
@@ -5,7 +5,7 @@
     "forceConsistentCasingInFileNames": true,
     "incremental": true,
     "isolatedModules": true,
-    "jsx": "preserve",
+    "jsx": "react-jsx",
```
Comment (nathggns) [inline site/tsconfig.json:8]: As above RE typescript/JSX.

**Review by nathggns — CHANGES_REQUESTED** [review]
I think there's something up with the JSX config changes, and also with the linter. It seems to be letting you do `React.*` even when `React` isn't in scope. Could you take a look?

File: package.json [hunk 15]
```diff
@@ -38,6 +38,9 @@
       "suppress-exit-code prettier --write"
     ]
   },
+  "resolutions": {
+    "@types/react": "^18.0.0"
+  },
```
[the remark this answers is not in the document]
Comment (CiaranMn) [reply package.json:43]: Updating `react-block-loader` is planned, yes

[the remark this answers is not in the document]
Comment (nonparibus) [reply package.json:43]: Internal ticket for this is at https://app.asana.com/0/1200211978612931/1202361504154770

File: packages/@blockprotocol/graph/README.md (hunk 1 above)
Comment (kachkaev) [reply packages/@blockprotocol/graph/README.md:63]: Thanks, fixed!

File: packages/@blockprotocol/core/tsconfig.json [hunk 16]
```diff
@@ -6,7 +6,7 @@
     "esModuleInterop": true,
     "forceConsistentCasingInFileNames": true,
     "isolatedModules": true,
-    "jsx": "preserve",
+    "jsx": "react-jsx",
```
[the remark this answers is not in the document]
Comment (kachkaev) [reply packages/@blockprotocol/core/tsconfig.json:9]: `tsc` is not used to build Next, but that’s not the case for some packages. Examples:

https://github.com/blockprotocol/blockprotocol/blob/cdb7805360d3412ea164ce9ff36c963f3a6df455/packages/mock-block-dock/package.json#L27-L30

My main goal of changing this field wast to get rid of `"jsx": "react"` (see other packages). That transform requires `import React from "react"` which we no longer have. JSX transform adds `import { jsx as _jsx } from "react/jsx-runtime"` automatically.

`"jsx": "preserve"` relies on downstream transpires (babel / swc) to do the job. This works for us, but may cause errors in other environments. AFAIU, sticking with `"jsx": "react-jsx"` is the safest bet in React 17+. If that’s not true, we can investigate further and create a follow-up PR.

File: packages/block-template/templates/custom-element/tsconfig.json (hunk 4 above)
Comment (kachkaev) [reply packages/block-template/templates/custom-element/tsconfig.json:7]: Replied in https://github.com/blockprotocol/blockprotocol/pull/441#r925489980

File: packages/block-template/templates/react/tsconfig.json (hunk 5 above)
Comment (kachkaev) [reply packages/block-template/templates/react/tsconfig.json:6]: Replied in https://github.com/blockprotocol/blockprotocol/pull/441#r925489980

File: packages/blockprotocol/tsconfig.json (hunk 6 above)
Comment (kachkaev) [reply packages/blockprotocol/tsconfig.json:5]: Replied in https://github.com/blockprotocol/blockprotocol/pull/441#r925489980

File: packages/mock-block-dock/dev/dev-app.tsx [hunk 17]
```diff
@@ -9,9 +9,9 @@ import { TestReactBlock } from "./test-react-block";
 
 const node = document.getElementById("app");
 
-const DevApp = () => {
-  const [entityId, setEntityId] = React.useState(1);
-  const [name, setName] = React.useState("World");
+const DevApp: FunctionComponent = () => {
```
[the remark this answers is not in the document]
Comment (kachkaev) [reply packages/mock-block-dock/dev/dev-app.tsx:12]: Yeah I agree that `FunctionComponent` does not add much here. I added this type for alignment with other components we have. In theory, we can run a [codemod](https://github.com/gndelia/codemod-replace-react-fc-typescript) and ditch FunctionComponent completely. We can discuss that separately ([internal slack thread](https://hashintel.slack.com/archives/C02TWBTT3ED/p1657623803157669)).

File: packages/block-template/templates/custom-element/src/dev.tsx (hunk 3 above)
Comment (kachkaev) [reply packages/block-template/templates/custom-element/src/dev.tsx]: Makes sense! Done: https://github.com/blockprotocol/blockprotocol/pull/441/commits/035d1e952ec284297aea8d2ff098c549679aca02

File: site/tsconfig.json (hunk 14 above)
Comment (kachkaev) [reply site/tsconfig.json:8]: Replied in https://github.com/blockprotocol/blockprotocol/pull/441#discussion_r925489980

File: site/src/components/modal/modal.tsx [hunk 18]
```diff
@@ -25,7 +24,7 @@ type ModalProps = MuiModalProps & {
   contentStyle?: SxProps<Theme>;
 };
 
-export const Modal: React.FC<ModalProps> = ({
+export const Modal: React.FunctionComponent<ModalProps> = ({
```
[the remark this answers is not in the document]
Comment (kachkaev) [reply site/src/components/modal/modal.tsx]: Thanks a lot for spotting this Nate! I fixed linting in https://github.com/blockprotocol/blockprotocol/pull/441/commits/fa4d077c035111b78641d5821276590a868fe8af and got rid of remaining `React.` cases in https://github.com/blockprotocol/blockprotocol/pull/441/commits/5fe0e729022e4dd51f8c7b18ec9d2b4ef65d8409.

File: packages/@blockprotocol/core/tsconfig.json (hunk 16 above)
[the remark this answers is not in the document]
Comment (nathggns) [reply packages/@blockprotocol/core/tsconfig.json:9]: > My main goal when changing this field was to get rid of "jsx": "react"

Totally support that – it's necessary for the purpose of this PR – removing `import React from "react";`. 

> "jsx": "preserve" relies on downstream transpires (babel / swc) to do the job. This works for us, but may cause errors in other environments

This pull request seems to be about changing imports, not about changing where our JSX is compiled. I don't really know the implications of doing it in TypeScript vs in other build tools, so not sure if I can comment on this. Generally I think PRs should be tightly scoped and this seems beyond the scope of the PR.

[the remark this answers is not in the document]
Comment (nathggns) [reply packages/@blockprotocol/core/tsconfig.json:9]: > Are you suggesting that we:
> 
> * replace `"jsx": "react"` with `"jsx": "react-jsx"` in two packages (`mock-block-dock` and `react-block-loader`)
> * keep `"jsx": "preserve"` in others to reduce diff?

Yes

**Comment by nathggns** [comment]
Looking at this again now

**Review by nathggns — APPROVED** [review]

**Pull request merged**